File: PopularMovies/src/main/java/com/mchapagai/movies/service/MovieService.java
package com.mchapagai.movies.service;

import com.mchapagai.core.response.common.ReviewListResponse;
import com.mchapagai.core.response.common.VideoListResponse;
import com.mchapagai.core.response.movies.MovieCreditResponse;
import com.mchapagai.core.response.movies.MovieDetailsResponse;
import com.mchapagai.core.response.movies.MovieListResponse;
import com.mchapagai.movies.model.binding.CombinedPersonResponse;
import com.mchapagai.movies.model.binding.PersonResponse;

import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Single;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface MovieService {

    @GET("discover/movie")
    Flowable<MovieListResponse> discoverMovies(@Query("page") int page,
                                               @Query("sort_by") String sortBy);

    @GET("movie/{id}/videos")
    Observable<VideoListResponse> getMovieVideosById(@Path("id") int movieId);

    @GET("movie/{id}/reviews")
    Observable<ReviewListResponse> getMovieReviewsById(@Path("id") int movieId);

    @GET("movie/{id}")
    Observable<MovieDetailsResponse> getMovieDetails(@Path("id") int movieId);

    @GET("movie/{id}/credits")
    Observable<MovieCreditResponse> getMovieCreditDetailsByCreditId(@Path("id") int movieId);

    @GET("person/{personId}")
    Single<PersonResponse> getPersonDetailsById(@Path("personId") int id);

    @GET("person/{personId}/combined_credits")
    Observable<CombinedPersonResponse> getPersonCombinedDetailsById(@Path("personId") int personId);

    @GET("search/movie")
    Observable<MovieListResponse> searchMovies(@Query("query") String q);

}

File: Core/src/main/java/com/mchapagai/core/service/MovieService.kt
package com.mchapagai.core.service

import com.mchapagai.core.response.common.ReviewListResponse
import com.mchapagai.core.response.common.VideoListResponse
import com.mchapagai.core.response.movies.MovieCreditResponse
import com.mchapagai.core.response.movies.MovieDetailsResponse
import com.mchapagai.core.response.movies.MovieListResponse
import io.reactivex.Flowable
import io.reactivex.Observable
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

interface MovieService {
    @GET("discover/movie")
    fun discoverMovies(
        @Query("page") page: Int,
        @Query("sort_by") sortBy: String
    ): Flowable<MovieListResponse>

    // append: credits,videos,videos,reviews,similar,external_id,images
    // @GET("tv/{tvId}?append_to_response=videos")
    @GET("movie/{movieId}")
    fun fetchMovieDetailsByMovieId(
        @Path("movieId") movieId: Int
    ): Observable<MovieDetailsResponse>

    @GET("movie/{movieId}/credits")
    fun getMovieCreditDetailsByCreditId(
        @Path("movieId") movieId: Int
    ): Observable<MovieCreditResponse>

    @GET("movie/{movieId}/videos")
    fun getMovieVideosById(
        @Path("movieId") movieId: Int
    ): Observable<VideoListResponse>

    @GET("movie/{movieId}/reviews")
    fun getMovieReviewsById(
        @Path("movieId") movieId: Int
    ): Observable<ReviewListResponse>
}
